lib/bolt/inventory/inventory2.rb

```diff
@@ -219,102 +197,202 @@ def expand_targets(targets)
+          add_to_group([t], 'all')
         end
+        t
       end
-      private :set_plugin_hooks
 
-      def add_target(current_group, target, desired_group, track = { 'all' => nil })
-        if current_group.name == desired_group
-          # Group to add to is found
-          t_name = target.name
-          # Add target to targets hash
-          target_hash = { 'name' => t_name }.merge(target.options)
-          target_hash['uri'] = target.uri if target.uri
-          current_group.targets[t_name] = target_hash
-
-          # Inherit facts, vars, and features from hierarchy
-          current_group_data = { facts: current_group.facts,
-                                 vars: current_group.vars,
-                                 features: current_group.features,
-                                 plugin_hooks: current_group.plugin_hooks }
-          data = inherit_data(track, current_group.name, current_group_data)
-          set_facts(t_name, @target_facts[t_name] ? data[:facts].merge(@target_facts[t_name]) : data[:facts])
-          set_vars_from_hash(t_name, @target_vars[t_name] ? data[:vars].merge(@target_vars[t_name]) : data[:vars])
-          data[:features].each do |feature|
-            set_feature(target, feature)
-          end
-          hook_data = @config.plugin_hooks.merge(data[:plugin_hooks])
-          hash = if @target_plugin_hooks[t_name]
-                   hook_data.merge(@target_plugin_hooks[t_name])
-                 else
-                   hook_data
-                 end
-          set_plugin_hooks(t_name, hash)
-          return true
+      # CODEREVIEW: This is helpful in group2 for validation. Where is a good place for it?
+      # Also is loading the addressable code here OK?
+      def self.parse_uri(string)
+        require 'addressable/uri'
+        if string.nil?
+          nil
+        # Forbid empty uri
+        elsif string.empty?
+          raise Bolt::ParseError, "Could not parse target URI: URI is empty string"
+        elsif string =~ %r{^[^:]+://}
+          Addressable::URI.parse(string)
+        else
+          # Initialize with an empty scheme to ensure we parse the hostname correctly
+          Addressable::URI.parse("//#{string}")
         end
-        # Recurse on children Groups if not desired_group
-        current_group.groups.each do |child_group|
-          track[child_group.name] = current_group
-          add_target(child_group, target, desired_group, track)
+      rescue Addressable::URI::InvalidURIError => e
+        raise Bolt::ParseError, "Could not parse target URI: #{e.message}"
+      end
+
+      def set_var(target, var_hash)
+        @targets[target.name]['vars'] = @targets[target.name]['vars'].merge(var_hash)
+        update_target(target)
+      end
+
+      def vars(target)
+        @targets[target.name]['cached_state']['vars'] || {}
+      end
+
+      def add_facts(target, new_facts = {})
+        # CODEREVIEW: Do we want to warn for set methods?
```

Comment:
I don't think there's a need to. The scenario where someone unintentionally has an empty hash where they intended to have data seem unlikely, and the puppet function doesn't allow this to be nil.